File: Recursion/bestSum.py
def bestSum(targetSum, array, memo={}):
  
  if targetSum in memo:
    return memo[targetSum]

  if targetSum == 0:
    return []

  if targetSum < 0: 
    return None

  bestCombo = None

  for num in array:
    remainder = targetSum - num
    result = bestSum( remainder, array )

    if result is not None:
      newestCombo = result + [num]
      if bestCombo is None or len(newestCombo) < len(bestCombo):
        bestCombo = newestCombo

  memo[targetSum] = bestCombo
  return bestCombo

print(bestSum(70, [5, 3, 4,1,1,1,1,1,1,1,1,1,1,1,1,1,1,30]))

File: Recursion/bestSum.js
let calls = 0

const bestSum = (target, numbers, memo={}) => {
  calls += 1
  if(target in memo) return memo[target]


  if( target === 0 ) return []

  if( target < 0 ) return null

  let bestCombo = null

  for(let num of numbers){

    const remainder = target - num
    
    const result = bestSum(remainder, numbers, memo)

    if(result){
      const newCombo = [...result, num]
      if(!bestCombo || newCombo.length < bestCombo.length){
        bestCombo = newCombo
      }
    }

  }

  memo[target] = bestCombo
  return bestCombo

}


console.log(bestSum(16, [1, 1, 1, 5, 8]), {calls})